### PR TITLE
Adding parallel compilation support for c++ modules

### PR DESCRIPTION
**Pull request by waruqi**
https://github.com/xmake-io/xmake/pull/2641

### CODE REVIEW

**Comment by Arthapz** [comment]
can you rebase batchjobs on current cxxmodules ? i want to try to do batchjobs for msvc

**Comment by waruqi** [comment]
Please wait a while, I'm working on improving rules, the current rule mechanism doesn't handle it very well at the moment.

**Comment by waruqi** [comment]
I have rebased it. I have currently got gcc/headerunits to implement parallelism, but build_modules hasn't had time to implement it yet.

File: xmake/rules/c++/modules/xmake.lua [hunk 1]
```diff
@@ -50,7 +50,6 @@ rule("c++.build.modules")
 
 -- build modules
 rule("c++.build.modules.builder")
-    set_sourcekinds("cxx")
```
Comment (Arthapz) [inline xmake/rules/c++/modules/xmake.lua:53]: How do we generate module dependencies from cpp files without this ?

Comment (waruqi) [reply xmake/rules/c++/modules/xmake.lua:53]: please wait, I'll improve it

Comment (waruqi) [reply xmake/rules/c++/modules/xmake.lua:53]: I have fixed it.

**Comment by waruqi** [comment]
It works fine for gcc/windows now, but there is still an issue for clang.

```lua
cd dependence
xmake clean -a; xmake f --toolchain=llvm -c;xmake -r -v -j1
/usr/local/Cellar/llvm/13.0.0_2/bin/clang -c -Qunused-arguments -m64 -isysroot /Applications/Xcode.app/Contents/Developer/Platfo
rms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk -std=c++20 -fmodules -fbuiltin-module-map -fimplicit-modules -fno-implicit-mod
ule-maps -fmodule-file=hello=build/.gens/dependence/macosx/x86_64/release/rules/modules/cache/hello.pcm -fmodule-file=mod=build/
.gens/dependence/macosx/x86_64/release/rules/modules/cache/mod.pcm -o build/.objs/dependence/macosx/x86_64/release/src/hello_imp
l.cpp.o src/hello_impl.cpp
checking for flags (-MMD -MF) ... ok
checking for flags (-fdiagnostics-color=always) ... ok
error: src/hello_impl.cpp:4:1: error: definition of module 'hello' is not available; use -fmodule-file= to specify path to preco
mpiled module interface
module hello;
```

but `-fmodule-file=hello=build/.gens/dependence/macosx/x86_64/release/rules/modules/cache/hello.pcm` exists

**Comment by waruqi** [comment]
> It works fine for gcc/windows now, but there is still an issue for clang.
> 
> ```lua
> cd dependence
> xmake clean -a; xmake f --toolchain=llvm -c;xmake -r -v -j1
> /usr/local/Cellar/llvm/13.0.0_2/bin/clang -c -Qunused-arguments -m64 -isysroot /Applications/Xcode.app/Contents/Developer/Platfo
> rms/MacOSX.platform/Developer/SDKs/MacOSX11.3.sdk -std=c++20 -fmodules -fbuiltin-module-map -fimplicit-modules -fno-implicit-mod
> ule-maps -fmodule-file=hello=build/.gens/dependence/macosx/x86_64/release/rules/modules/cache/hello.pcm -fmodule-file=mod=build/
> .gens/dependence/macosx/x86_64/release/rules/modules/cache/mod.pcm -o build/.objs/dependence/macosx/x86_64/release/src/hello_imp
> l.cpp.o src/hello_impl.cpp
> checking for flags (-MMD -MF) ... ok
> checking for flags (-fdiagnostics-color=always) ... ok
> error: src/hello_impl.cpp:4:1: error: definition of module 'hello' is not available; use -fmodule-file= to specify path to preco
> mpiled module interface
> module hello;
> ```
> 
> but `-fmodule-file=hello=build/.gens/dependence/macosx/x86_64/release/rules/modules/cache/hello.pcm` exists

batchjobs works fine now, this issue should not be related to batchjobs, the original branch of cxxmodules also has this issue

**Pull request merged**